File: java/src/main/java/section2/function/Function.java
package section2.function;

public class Function {

    public static void main(String[] args) {
        printAll("A", "B", "C");
        String[] strings = {"A", "B", "C"};
        printAll(strings);
    }

    public int max(int a, int b) {
        if (a > b) {
            return a;
        }
        return b;
    }

    public void repeat(String str, int num, boolean useNewLine) {
        for (int i = 1; i <= num; i++) {
            if (useNewLine) {
                System.out.println(str);
            } else {
                System.out.print(str);
            }
        }
    }

    public void repeat(String str, int num) { // overloading
        repeat(str, num, true);
    }

    public void repeat(String str) { // overloading
        repeat(str, 3, true);
    }

    public static void printAll(String... strings) {
        for (String str : strings) {
            System.out.println(str);
        }
    }

}

File: kotlin/src/main/kotlin/section2/function/Function.kt
package section2.function

fun main() {
    repeat("Hello", 5)
    repeat("World", useNewLine = false) // named argument

    printNameAndGender("남성", "김영재")
    printNameAndGender(
        gender = "남성",
        name = "김영재"
    ) // named argument

    // 가변인자
    printAll("A", "B", "C") // 콤마로
    val array = arrayOf("A", "B", "C")
    printAll(*array) // spread operator. 배열을 풀어서 전달
    printAll(*array, "D", "E")
}

fun max(a: Int, b: Int) =
    if (a > b) a else b

fun repeat(
    str: String,
    num: Int = 3,
    useNewLine: Boolean = true
) {
    for (i in 1..num) {
        print(str)
        if (useNewLine) {
            println()
        }
    }
}

fun printNameAndGender(name: String, gender: String) { // 동일한 타입이 연속으로 존재
    println(name)
    println(gender)
}

fun printAll(vararg strings: String) {
    for (str in strings) {
        println(str)
    }
}
